src/protocol/libp2p/kademlia/store.rs

```diff
@@ -32,28 +32,184 @@ pub enum MemoryStoreEvent {}
 pub struct MemoryStore {
     /// Records.
     records: HashMap<Key, Record>,
+    /// Configuration.
+    config: MemoryStoreConfig,
 }
 
 impl MemoryStore {
     /// Create new [`MemoryStore`].
     pub fn new() -> Self {
         Self {
             records: HashMap::new(),
+            config: MemoryStoreConfig::default(),
+        }
+    }
+
+    /// Create new [`MemoryStore`] with the provided configuration.
+    pub fn with_config(config: MemoryStoreConfig) -> Self {
+        Self {
+            records: HashMap::new(),
+            config,
         }
     }
 
     /// Try to get record from local store for `key`.
-    pub fn get(&self, key: &Key) -> Option<&Record> {
-        self.records.get(key)
+    pub fn get(&mut self, key: &Key) -> Option<&Record> {
+        let is_expired = self
+            .records
+            .get(key)
+            .map_or(false, |record| record.is_expired(std::time::Instant::now()));
+
+        if is_expired {
+            self.records.remove(key);
+            None
+        } else {
+            self.records.get(key)
+        }
     }
 
     /// Store record.
     pub fn put(&mut self, record: Record) {
-        self.records.insert(record.key.clone(), record);
+        if record.value.len() >= self.config.max_record_size_bytes {
+            return;
+        }
+
+        let len = self.records.len();
+        match self.records.entry(record.key.clone()) {
+            Entry::Occupied(mut entry) => {
+                // Lean towards the new record.
+                if let (Some(stored_record_ttl), Some(new_record_ttl)) =
+                    (entry.get().expires, record.expires)
+                {
+                    if stored_record_ttl > new_record_ttl {
+                        return;
+                    }
+                }
+
+                entry.insert(record);
+            }
+
+            Entry::Vacant(entry) => {
+                if len >= self.config.max_records {
+                    return;
```

Comment:
May be try cleaning up free space here by removing expired records? Out of scope of this PR, as this would require storing records both sorted by their TTL & hashed by their key. But something to consider as a low-priority follow up.